### DISCUSSION TITLE
Adding global tags for metrics

### DISCUSSION THREAD

**Discussion by EugeneKrapivin** [open-telemetry/opentelemetry-dotnet#4420] [General]
I'm currently at the stage of finding all the missing pieces to have a comparable reporting "feeling" we are currently having.

Currently all metrics reported from our services have some default tags/dimensions, like: dc, environment, service name. We aren't recording them manually every time we record a metric, but rather set those values as defaults at setup time. this really help since when querying we sometimes have metrics from multiple services, multiple service versions, multiple environments and even at time multiple datacenters in the same TSDB.

how would we go around adding those tags to all metrics exported from the service? I thought that adding those as attributes to the resource would do the trick
```csharp
metricsBuilder.ConfigureResource(resourceBuilder =>
{
    resourceBuilder
        .AddService("ClickService")
        .AddAttributes(new[]
        {
            new KeyValuePair<string, object>("dc", Environment.GetEnvironmentVariable("dc") ?? "us1"),
            new KeyValuePair<string, object>("env", Environment.GetEnvironmentVariable("env") ?? "local"),
        })
        .AddTelemetrySdk();
});
```
however, I was wrong.
I would assume that there would be a way to set default tags on a `Meter` so that the process of reporting default tags is not repetitive on each `Add`/`Record` call.

**Comment by cijothomas** [comment]
Resource is what you need to model environments, service name etc. Is that not working for you? Could you elaborate more on why is it not working?

**Reply by EugeneKrapivin** [reply]
Hello @cijothomas !
yeah, so I've expected these attributes I've set up on the resource builder would propagate to all of the exported metrics. However, they were not.

the full code (however, not working due to limitations of dotnetfiddle could be found here https://dotnetfiddle.net/aIRBu0
but I'll copy paste the important bits.

As you could see, I've added a resources, a service and added 2 attributes that I've pulled from my environment variables (with defaults in case they aren't defined).
```csharp
var host = Host
    .CreateDefaultBuilder()
    .ConfigureLogging(logger => logger.AddConsole())
    .ConfigureServices(services =>
    {
        services.AddOpenTelemetry()
            .WithMetrics(metricsBuilder =>
            {
                metricsBuilder.ConfigureResource(resourceBuilder =>
                {
                    resourceBuilder
                        .AddService("ClickService")
                        .AddAttributes(new[]
                        {
                            new KeyValuePair<string, object>("dc", Environment.GetEnvironmentVariable("dc") ?? "us1"),
                            new KeyValuePair<string, object>("env", Environment.GetEnvironmentVariable("env") ?? "local"),
                        })
                        .AddTelemetrySdk();
                });
                var meterProviderBuilder = metricsBuilder.AddMeter("custom-console-meter");

                metricsBuilder.AddConsoleExporter();

                metricsBuilder.AddPrometheusHttpListener(httpListenerOptions =>
                {
                    httpListenerOptions.UriPrefixes = new[] { "http://localhost:9464/" };
                });
            });

        services.AddHostedService<ClickAppHost>();
    });

await host.RunConsoleAsync();
```

In the actual code I'm doing:
```csharp
Console.WriteLine("Click a key");
var sw = Stopwatch.GetTimestamp();

var key = Console.ReadKey(intercept: true);

var elapsed = Stopwatch.GetElapsedTime(sw);
var combination = GetKeyLabel(key);
Console.WriteLine($"Clicked: {combination}");

Meters.ClickInstrument.Add(1, new KeyValuePair<string, object?>("key", combination));
Meters.DurationInstrumentation.Record((long)elapsed.TotalMilliseconds);
```

I was under the assumption that both metrics would be exported with `dc` and `env` as tags. However, getting the prometheus endpoint returns only the tags defined directly on the `Record`/`Add` methods:
```
# TYPE click_c_clicks_total counter
# UNIT click_c_clicks_total clicks_total
# HELP click_c_clicks_total count the number of clicks
click_c_clicks_total{key="a"} 14 1682246108134
click_c_clicks_total{key="s"} 15 1682246108134
click_c_clicks_total{key="d"} 14 1682246108134
click_c_clicks_total{key="e"} 2 1682246108134
click_c_clicks_total{key="r"} 2 1682246108134
click_c_clicks_total{key="w"} 1 1682246108134

# TYPE click_t_ms histogram
# UNIT click_t_ms ms
# HELP click_t_ms time to click after prompted
click_t_ms_bucket{le="0"} 0 1682246108134
click_t_ms_bucket{le="5"} 0 1682246108134
click_t_ms_bucket{le="10"} 0 1682246108134
click_t_ms_bucket{le="25"} 0 1682246108134
click_t_ms_bucket{le="50"} 0 1682246108134
click_t_ms_bucket{le="75"} 17 1682246108134
click_t_ms_bucket{le="100"} 24 1682246108134
click_t_ms_bucket{le="250"} 42 1682246108134
click_t_ms_bucket{le="500"} 44 1682246108134
click_t_ms_bucket{le="750"} 47 1682246108134
click_t_ms_bucket{le="1000"} 47 1682246108134
click_t_ms_bucket{le="2500"} 47 1682246108134
click_t_ms_bucket{le="5000"} 47 1682246108134
click_t_ms_bucket{le="7500"} 47 1682246108134
click_t_ms_bucket{le="10000"} 47 1682246108134
click_t_ms_bucket{le="+Inf"} 48 1682246108134
click_t_ms_sum 29085 1682246108134
click_t_ms_count 48 1682246108134

# EOF
```

**Reply by cijothomas** [reply]
> I was under the assumption that both metrics would be exported with dc and env as tags

That is not correct. Resources are separate from Tags/Attributes. Backends may choose to convert Resources as attributes, but this is up to individual backends.
The current prometheus exporter does not process Resources at all. There is an experimental spec written for it (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#resource-attributes-1), but not yet implemented here.

**Reply by EugeneKrapivin** [reply]
I see. 
Thanks for the reference.
Would would working with an OTel collector which will get the metrics into Prometheus be a valid workaround?

**Reply by cijothomas** [reply]
Yes, OTel Collector's Prometheus Exporter has the ability to convert Resource as Metric Labels.
https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusexporter#getting-started (`resource_to_telemetry_conversion`)

**Reply by EugeneKrapivin** [reply]
After researching a bit further and having some time to pounder, I still feel that there is a gap between the spec and real world use.
For example: I have a system with a couple of hundreds of services, each having multiple instances, some deployed in k8s and some are still deployed as processes in nomad. It is a common use-case to have my services report as part of the metric labels information like the dc, env, region, service name, service version so that we could easily filter and query the data, especially when debugging an issue in real-time. This is even more prominent when using contrib instrumentations, for which I have no control over labels. 

What would be the suggested methodology to achieve that at this point in time?

From reading the spec a bit "Producers SHOULD prevent the presence of multiple Metric identities for a given name with the same **Resource** and Scope **attributes**." (source: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#opentelemetry-protocol-data-model-producer-recommendations). Doesn't that imply that metrics aggregation (and reporting I would assume) should take in accounts the resource attributes?

Thanks,
Eugene.

**Reply by cijothomas** [reply]
> After researching a bit further and having some time to pounder, I still feel that there is a gap between the spec and real world use.
> For example: I have a system with a couple of hundreds of services, each having multiple instances, some deployed in k8s and some are still deployed as processes in nomad. It is a common use-case to have my services report as part of the metric labels information like the **dc, env, region, service name, service version** so that we could easily filter and query the data, especially when debugging an issue in real-time. This is even more prominent when using contrib instrumentations, for which I have no control over labels.
> 
> What would be the suggested methodology to achieve that at this point in time?

**These** should be modelled as Resource by the application owner. It should not be reported as attributes/tags by any instrumentation library.

**Reply by cijothomas** [reply]
> Doesn't that imply that metrics aggregation (and reporting I would assume) should take in accounts the resource attributes?

No.

**Reply by EugeneKrapivin** [reply]
> These should be modelled as Resource by the application owner. It should not be reported as attributes/tags by any instrumentation library.

However, resource Attributes and Service information is not passed along (at least to Prometheus) making them a bit redundant, since they cannot be really used to filter data points. What are the use of those Resource, Service and attribute definitions if they aren't externalized by the Exporter?

**Reply by cijothomas** [reply]
> What are the use of those Resource, Service and attribute definitions if they aren't externalized by the Exporter?

Its an exporter issue only and must be fixed by the exporter itself.
The official OTLP Exporter supports it.
The official Prometheus exporter would also support it. (The spec is experimental for that).
For any other exporter, please file an issue with the exporter authors.

**Comment by jaaywags** [comment]
I found [this solution](https://stackoverflow.com/a/71127386/3543339) to be what I needed.

I had dotnet sending attributes the same exact way, but

> The resource attributes to metrics label conversion is disabled by default

In your open telemetry collector config file, you should have an exporter for prometheus. Update it to look like this.

```yaml
exporters:
    prometheus:
      endpoint: "0.0.0.0:8889"
      resource_to_telemetry_conversion:
        enabled: true
```

**Reply by cijothomas** [reply]
https://prometheus.io/docs/guides/opentelemetry/#promoting-resource-attributes 
Prometheus natively supports Resource now as well.
